example/synchronous_resnet/resnet_controller.py

```diff
@@ -0,0 +1,150 @@
+            self.valid_history_errs[self._epoch][worker_id] = req_info['valid_err']
+            self.test_history_errs[self._epoch][worker_id] = req_info['test_err']
+
+            if not any([i is None for i in self.valid_history_errs[self._epoch]]):
+                print('Epoch %d is done'%req_info['epoch'])
+                valid_err = sum(self.valid_history_errs[self._epoch]) / float(self.nb_worker)
+
+                if valid_err <= self.best_dict['best_valid']:
+                    self.best_dict['best_epoch'] = self._epoch
+                    self.best_dict['best_valid'] = valid_err
+                    self.bad_counter = 0
+                    control_response = 'best'
+                    print("Best error valid:", valid_err)
+                else:
+                    self.bad_counter += 1
+                self.valid_history_errs += [[None for i in range(self.nb_worker)]]
+                self.test_history_errs += [[None for i in range(self.nb_worker)]]
+                self._epoch += 1
+
+        elif req == 'splits':
+            # the controller never loads the dataset but the worker doesn't
+            # know how many workers there are
+            train_len = req_info['train_len'] // self.nb_worker
+            valid_len = req_info['valid_len'] // self.nb_worker
+            test_len = req_info['test_len'] // self.nb_worker
+            splits = dict(train_splits=[train_len * worker_id, train_len * (worker_id + 1)],
+                          valid_splits=[valid_len * worker_id, valid_len * (worker_id + 1)],
+                          test_splits=[test_len * worker_id, test_len * (worker_id + 1)])
+            control_response = splits
+
+            # kind of when the training start but not really
+            self.start_time = time.time()
+
+        elif req == 'seed':
+            control_response = self.seed
+
+        if self.bad_counter > self.patience:
+            print("Early stopping!")
+            end_time = time.time() - self.start_time
+            # should terminate with best printing and best dumping of params
+            # and then close everything
+            print("Best error valid:", self.best_dict['best_valid'])
+            test_err = sum(self.test_history_errs[self.best_dict['best_epoch']]) / \
+                    float(self.nb_worker)
+            print("Best error test:", test_err)
+            print( ("Training took %.1fs" % (end_time)), file=sys.stderr)
+            control_response = 'stop'
+
+        return control_response
+
+
+def resnet_control(saveFreq=1110, saveto=None):
+    parser = Controller.default_parser()
+    parser.add_argument('--seed', default=1234, type=int,
+                        required=False, help='Maximum mini-batches to train upon in total.')
+    parser.add_argument('--patience', default=10, type=int, required=False,
+                        help='Maximum patience when failing to get better validation results.')
```

Comment:
From the original lstm example it is from validation epoch actually, so since the validation is not done after every training epoch then this number has to be not too large. Could make the help more useful indeed.